engine/table/src/main/java/io/deephaven/engine/table/impl/multijoin/StaticMultiJoinStateManagerTypedBase.java

```diff
@@ -0,0 +1,218 @@
+/**
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.engine.table.impl.multijoin;
+
+import io.deephaven.base.verify.Require;
+import io.deephaven.chunk.Chunk;
+import io.deephaven.chunk.ChunkType;
+import io.deephaven.chunk.attributes.Values;
+import io.deephaven.engine.rowset.RowSequence;
+import io.deephaven.engine.rowset.RowSet;
+import io.deephaven.engine.table.ColumnSource;
+import io.deephaven.engine.table.Table;
+import io.deephaven.engine.table.WritableColumnSource;
+import io.deephaven.engine.table.impl.MultiJoinStateManager;
+import io.deephaven.engine.table.impl.sources.ArrayBackedColumnSource;
+import io.deephaven.engine.table.impl.sources.InMemoryColumnSource;
+import io.deephaven.engine.table.impl.sources.LongArraySource;
+import io.deephaven.engine.table.impl.sources.immutable.ImmutableLongArraySource;
+import io.deephaven.engine.table.impl.util.ChunkUtils;
+import io.deephaven.engine.table.impl.util.LongColumnSourceWritableRowRedirection;
+import io.deephaven.engine.table.impl.util.TypedHasherUtil;
+import io.deephaven.engine.table.impl.util.TypedHasherUtil.BuildOrProbeContext.BuildContext;
+import io.deephaven.engine.table.impl.util.WritableRowRedirection;
+import io.deephaven.util.QueryConstants;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static io.deephaven.engine.table.impl.JoinControl.CHUNK_SIZE;
+import static io.deephaven.engine.table.impl.JoinControl.MAX_TABLE_SIZE;
+import static io.deephaven.engine.table.impl.util.TypedHasherUtil.getKeyChunks;
+
+public abstract class StaticMultiJoinStateManagerTypedBase implements MultiJoinStateManager {
+    protected final ColumnSource<?>[] keySourcesForErrorMessages;
+    private final List<LongArraySource> indexSources = new ArrayList<>();
+
+    public static final long NO_RIGHT_STATE_VALUE = RowSet.NULL_ROW_KEY;
+    public static final long EMPTY_RIGHT_STATE = QueryConstants.NULL_LONG;
+    public static final long DUPLICATE_RIGHT_STATE = -2;
+
+    // The number of slots in our hash table.
+    protected int tableSize;
+
+    // The number of entries in our hash table in use.
+    protected long numEntries = 0;
+
+    // The table will be rehashed to a load factor of targetLoadFactor if our loadFactor exceeds maximumLoadFactor
+    // or if it falls below minimum load factor we will instead contract the table.
+    private final double maximumLoadFactor;
+
+    // The keys for our hash entries.
+    protected final ChunkType[] chunkTypes;
+    protected final WritableColumnSource[] mainKeySources;
+
+    // The output sources representing the keys of our joined table.
+    protected final WritableColumnSource[] outputKeySources;
+
+    // Store sentinel information and maps hash slots to output row keys.
+    protected ImmutableLongArraySource slotToOutputRow = new ImmutableLongArraySource();
```

Comment:
We seem to have a conflict here. This is an immutable, so we cannot `ensureCapacity` on it more than once, but we're doing so in the constructor based on what the `JoinControl` says, rather than doing it at the end when building our output table.

Reply:
This has to be done initially to set the size of the immutable columns.  After this, the columns are resized during the rehash.

Reply:
I think I understand. We should just be sure to avoid re-ensuring capacity.